network/connPerfMon.go

```diff
@@ -0,0 +1,351 @@
+			} else {
+				pm.connectionDelay[peer] = 0
+			}
+		}
+		return
+	}
+	pm.lastIncomingMsgTime = msg.Received
+	// otherwise, once we recieved a message from each of the peers, move to the sync stage.
+	pm.advanceStage(pmStageSync, msg.Received)
+}
+
+// notifySync waits for all the peers connection's to go into an idle phase.
+// when we go into this stage, the lastPeerMsgTime will be already updated
+// with the recent message time per peer.
+func (pm *connectionPerformanceMonitor) notifySync(msg *IncomingMessage) {
+	minMsgInterval := pm.updateMessageIdlingInterval(msg.Received)
+	if minMsgInterval > int64(pmSyncIdleTime) || (msg.Received-pm.stageStartTime > int64(pmSyncMaxTime)) {
+		// if we hit the first expression, then it means that we've managed to sync up the connections.
+		// otherwise, we've failed to sync up the connections. That's not great, as we're likly to
+		// have some "penelties" applied, but we can't do much about it.
+		pm.accumulateMessage(msg, true)
+		pm.advanceStage(pmStageAccumulate, msg.Received)
+	}
+}
+
+// notifyAccumulate accumulate the incoming message as needed, and waiting between pm.accumulationTime to
+// (pm.accumulationTime + pmAccumulationTimeRange) before moving to the next stage.
+func (pm *connectionPerformanceMonitor) notifyAccumulate(msg *IncomingMessage) {
+	minMsgInterval := pm.updateMessageIdlingInterval(msg.Received)
+	if msg.Received-pm.stageStartTime >= pm.accumulationTime {
+		if minMsgInterval > int64(pmAccumulationIdlingTime) ||
+			(msg.Received-pm.stageStartTime >= pm.accumulationTime+int64(pmAccumulationTimeRange)) {
+			// move to the next stage.
+			pm.advanceStage(pmStageStopping, msg.Received)
+			return
+		}
+	}
+	pm.accumulateMessage(msg, true)
+	pm.pruneOldMessages(msg.Received)
+}
+
+// notifyStopping attempts to stop the message accumulation. Once we reach this stage, no new messages are being
+// added, and old pending messages are being pruned. Once all messages are pruned, it moves to the next stage.
+func (pm *connectionPerformanceMonitor) notifyStopping(msg *IncomingMessage) {
+	pm.accumulateMessage(msg, false)
+	pm.pruneOldMessages(msg.Received)
+	if len(pm.pendingMessages) > 0 {
+		return
+	}
+	// time to wrap up.
+	if pm.msgCount > 0 {
+		for peer := range pm.monitoredConnections {
+			pm.connectionDelay[peer] /= int64(pm.msgCount)
+		}
+	}
+	pm.advanceStage(pmStageStopped, msg.Received)
+}
+
+// advanceStage set the stage variable and update the stage start time.
+func (pm *connectionPerformanceMonitor) advanceStage(newStage pmStage, now int64) {
+	pm.stage = newStage
+	pm.stageStartTime = now
+}
+
+// updateMessageIdlingInterval updates the last message received timestamps and determines how long it has been since
+// the last message was received on any of the incoming peers
+func (pm *connectionPerformanceMonitor) updateMessageIdlingInterval(now int64) (minMsgInterval int64) {
+	currentIncomingMsgTime := pm.lastIncomingMsgTime
```

Comment:
should not lastIncomingMsgTime be atomic?

Reply:
All access to lastIncomingMsgTime is behind the connectionPerformanceMonitor.lock, so I don't think it's needed.